### PR TITLE
Add get_mime_data(), set_mime_data() and supports_mime_type()  to cario surface

### PR DESCRIPTION
**Pull request by jsparber**

### CODE REVIEW

File: src/surface.rs [hunk 1]
```diff
@@ -47,6 +47,61 @@ impl Surface {
                                                                   height))
         }
     }
+
+    pub fn get_mime_data(&self, mime_type: &str) -> Option<&[u8]> {
```
Comment (sdroege) [inline src/surface.rs]: This is unsafe: nothing is preventing the caller to set a different data for the mimetype after calling this function, and then using the (now freed) byte array of the previously set data

File: src/surface.rs [hunk 2]
```diff
@@ -47,6 +47,61 @@ impl Surface {
...
+
+        let status = unsafe {
+            ffi::cairo_surface_set_mime_data(
+                self.to_raw_none(),
+                mime_type.to_glib_none().0,
+                data,
+                size as u64,
+                Some(unbox::<T>),
+                user_data as *mut _,
+                )
```
Comment (EPashkin) [inline src/surface.rs]: Please make indent smaller

File: src/surface.rs (hunk 1 above)
Comment (jsparber) [reply src/surface.rs]: Should i make a copy of the data and return a vector?

Comment (sdroege) [reply src/surface.rs]: That would be one option, yes. Alternative would be something like the data getter from the image surface but that's very limiting (requires refcount of 1).

File: src/surface.rs (hunk 2 above)
Comment (jsparber) [reply src/surface.rs]: How wide should the indent be? Isn't it 4 spaces?

File: src/surface.rs (hunk 1 above)
Comment (jsparber) [reply src/surface.rs]: What would be the advantage of using a getter similar to the that one in image surface?

Comment (sdroege) [reply src/surface.rs]: No copying of the data into a `Vec` :)

Comment (jsparber) [reply src/surface.rs]: Oh I thought something more than that. I used to_vec()

Comment (sdroege) [reply src/surface.rs]: Great, thanks :)

File: src/surface.rs (hunk 2 above)
Comment (sdroege) [reply src/surface.rs]: 4 spaces, yes

File: src/surface.rs [hunk 3]
```diff
@@ -47,6 +47,60 @@ impl Surface {
...
+                ).to_vec())
+            } else {
+                None
+            }
+        }
+    }
+
+    pub fn set_mime_data<T: AsRef<[u8]> + Send + 'static>(&self,
+                                                          mime_type: &str,
+                                                          slice: T) -> Result<(), Status> {
```
Comment (sdroege) [inline src/surface.rs]: This one should probably be indented a bit differntly:

```rust
    pub fn set_mime_data<T: AsRef<[u8]> + Send + 'static>(
        &self,
        mime_type: &str,
        slice: T) -> Result<(), Status>
```

Or something like that. Indent logically, not visually :)

File: src/surface.rs [hunk 4]
```diff
@@ -47,6 +47,60 @@ impl Surface {
...
+        let user_data = Box::into_raw(b);
+
+        let status = unsafe {
+            ffi::cairo_surface_set_mime_data(
+                self.to_raw_none(),
+                mime_type.to_glib_none().0,
+                data,
+                size as u64,
+                Some(unbox::<T>),
+                user_data as *mut _)
```
Comment (sdroege) [inline src/surface.rs]: The closing parenthesis should be on the next line at the same level as `ffi::`

File: src/surface.rs (hunk 2 above)
Comment (sdroege) [reply src/surface.rs]: That's why we really need rustfmt ;)

Comment (jsparber) [reply src/surface.rs]: i really want rustfmt.

**Comment by jsparber** [comment]
what should i use instead of to_glib_none() and ptr::null_mut(), since the CI fails because of them?

**Comment by jsparber** [comment]
I think i fixed all indentation issues, but we really need rustfmt

**Comment by sdroege** [comment]
> what should i use instead of to_glib_none() and ptr::null_mut(), since the CI fails because of them?

The former is only available if cairo is built with glib support. You should use `CStr` / `CString` for strings, or where are the problems?

For `ptr`, that should work. How does it fail?

**Comment by jsparber** [comment]
I didn't include ptr in surface, I guess cagro just got it from some where else. Anyhow it is fixed now. Also the CString thing.

**Comment by jsparber** [comment]
@sdroege std::ptr is included already inside the gvalue_impl macro, but this is used only when glib is enabled. How should I import in surface.rs?

**Comment by sdroege** [comment]
> @sdroege std::ptr is included already inside the gvalue_impl macro, but this is used only when glib is enabled therefore the build fails if glib is disabled. How should I import in surface.rs?

Just add a use statement for it? :)

**Comment by jsparber** [comment]
> > @sdroege std::ptr is included already inside the gvalue_impl macro, but this is used only when glib is enabled therefore the build fails if glib is disabled. How should I import in surface.rs?
> 
> Just add a use statement for it? :)

Rust doesn't like to include it twice. I now added an annotation to the use statement to include std::ptr only when glib is disabled.

**Comment by sdroege** [comment]
> Rust doesn't like to include it twice. I now added an annotation to the use statement to include std::ptr only when glib is disabled.

You could remove the use statement from the macro (and use it as `::std::ptr::...`).

File: src/surface.rs (hunk 2 above)
Comment (EPashkin) [reply src/surface.rs]: First comment was about closing bracket has same indent as parameters.

**Comment by jsparber** [comment]
> > Rust doesn't like to include it twice. I now added an annotation to the use statement to include std::ptr only when glib is disabled.
> 
> You could remove the use statement from the macro (and use it as `::std::ptr::...`).

i tried std::ptr before (which was wrong). let's hope now it works

File: src/surface.rs (hunk 2 above)
Comment (jsparber) [reply src/surface.rs]: I hope you are happy with the last change @EPashkin ;)

Comment (EPashkin) [reply src/surface.rs]: Yes, thanks 😉

File: src/lib.rs [hunk 5]
```diff
@@ -141,6 +140,7 @@ pub use xcb::{
 };
 
 pub mod prelude;
+pub mod constants;
```
Comment (sdroege) [inline src/lib.rs]: Maybe
```rust
mod constants;
pub use constants::*;
```
instead?

**Comment by sdroege** [comment]
Looks good to me, but you might want to add an unsafe getter for the data too that does not copy. In librsvg you can ensure that the data stays valid all the time, and could prevent expensive copies that way.

**Comment by jsparber** [comment]
just a unsave function which returns option<&[u8]>? Any ideas how i should name it? I thought about get_mime_data_raw()

File: src/lib.rs (hunk 5 above)
Comment (jsparber) [reply src/lib.rs]: Out of curiosity whats the difference between those two? Yours looks cleaner ;)

Comment (sdroege) [reply src/lib.rs]: Yours requires users to do `cairo::constants::BLABLA`, mine would be `cairo::BLABLA`

**Comment by sdroege** [comment]
> just a unsave function which returns option<&[u8]>? Any ideas how i should name it? I thought about get_mime_data_raw()

Sounds good to me

**Comment by sdroege** [comment]
Seems good to me now

**Comment by GuillaumeGomez** [comment]
Thanks!

**Pull request merged**